<launch>
    
    <!-- from bag: rosbag record /stereo_camera/left/image_rect/compressed /stereo_camera/right/image_rect/compressed /stereo_camera/left/camera_info /stereo_camera/right/camera_info /stereo_camera/depth/compressedDepth -->
    
    <param name="use_sim_time" type="bool" value="True"/>
  	
  	<arg name="pi/2" value="1.5707963267948966" />
    <arg name="optical_rotate" value="0 0 0 -$(arg pi/2) 0 -$(arg pi/2)" />
  	<node pkg="tf" type="static_transform_publisher" name="camera_base_link"
        args="$(arg optical_rotate) base_link stereo_camera 100" />  
  	
	<!-- Odometry: Choose between viso2_ros, fovis_ros and homemade approach (default) -->
	<node pkg="viso2_ros" type="stereo_odometer" name="stereo_odometer" output="screen">
		<remap from="stereo" to="/stereo_camera"/>
		<remap from="image" to="image_rect"/>
		<param name="base_link_frame_id" value="base_link"/>
		<param name="odom_frame_id" value="/odom"/>
		<param name="ref_frame_change_method" value="1"/>
		<param name="image_transport" type="string" value="compressed"/>
	</node>
	
	<!--
	<node pkg="fovis_ros" type="fovis_stereo_odometer" name="stereo_odometer" >
	    <remap from="/stereo/left/image" to="/stereo_camera/left/image_rect" />
	    <remap from="/stereo/right/image" to="/stereo_camera/right/image_rect" />
	    <remap from="/stereo/left/camera_info" to="/stereo_camera/left/camera_info" />
	    <remap from="/stereo/right/camera_info" to="/stereo_camera/right/camera_info" />
	    <remap from="odometry" to="stereo_odometer/odometry" />
    </node>
    -->	
    <!--
    <node pkg="rtabmap" type="stereo_odometry" name="stereo_odometer" output="screen">
  	  	<remap from="left/image_rect"       to="/stereo_camera/left/image_rect"/>
  	  	<remap from="right/image_rect"      to="/stereo_camera/right/image_rect"/>
  	  	<remap from="left/camera_info"      to="/stereo_camera/left/camera_info"/>
  	  	<remap from="right/camera_info"     to="/stereo_camera/right/camera_info"/>
	  	<remap from="odom"                  to="/stereo_odometer/odometry"/>
	  	<remap from="odom_depth"            to="/stereo_camera/depth2"/>
	  	
	  	<param name="left/image_transport" type="string" value="compressed"/>
	  	<param name="right/image_transport" type="string" value="compressed"/>
    
  	  	<param name="frame_id" type="string" value="base_link"/>
  	  	<param name="approx_sync" type="bool" value="false"/>

		<param name="Odom/Strategy" type="string" value="1"/>            
		<param name="Odom/RoiRatios" type="string" value="0.03 0.03 0.04 0.04"/>
		<param name="Odom/MaxDepth" type="string" value="10"/>
  	</node>
  	-->
      
  <group ns="rtabmap">	
    <!-- Visual SLAM (robot side) -->
    <!-- args: "delete_db_on_start" and "udebug" -->
    <node name="rtabmap" pkg="rtabmap" type="rtabmap" output="screen" args="--delete_db_on_start">
      <param name="subscribe_depth" type="bool" value="true"/>
      <param name="subscribe_laserScan" type="bool" value="false"/>
	
      <remap from="rgb/image" to="/stereo_camera/left/image_rect"/>
      <remap from="rgb/camera_info" to="/stereo_camera/left/camera_info"/>
      <remap from="depth/image" to="/stereo_camera/depth"/>
      <remap from="odom" to="/stereo_odometer/odometry"/>
      
      <param name="rgb/image_transport" type="string" value="compressed"/>
      <param name="depth/image_transport" type="string" value="compressedDepth"/>      
      
      <param name="frame_id" type="string" value="base_link"/>
      <param name="queue_size" type="int" value="30"/>
      
      <param name="RGBD/ToroIterations" type="string" value="0"/>
      <param name="Rtabmap/TimeThr" type="string" value="700"/>
      <param name="Rtabmap/DetectionRate" type="string" value="1"/>
      <param name="Kp/DetectorStrategy" type="string" value="0"/>
      <param name="Kp/WordsPerImage" type="string" value="200"/>
      <param name="Kp/RoiRatios" type="string" value="0.03 0.03 0.04 0.04"/>
      <param name="NN/NNStrategy" type="string" value="1"/>
      
      <param name="LccBow/MinInliers" type="string" value="10"/>
      <param name="LccBow/InlierDistance" type="string" value="0.02"/>
      <param name="LccBow/MaxDepth" type="string" value="3"/>
      
      <param name="LccReextract/FeatureType" type="string" value="4"/>
	  <param name="LccReextract/Activated" type="string" value="true"/>
	  <param name="LccReextract/MaxDepth" type="string" value="10"/>
	  <param name="LccReextract/NNDR" type="string" value="0.8"/>
	  <param name="LccReextract/NNType" type="string" value="3"/> 	  	
    </node>
  
    <!-- Visualisation (client side) -->
    <!--
    <node pkg="rtabmap" type="rtabmapviz" name="rtabmapviz" args="-d $(find rtabmap)/launch/config/rgbd_gui.ini" output="screen">
      <param name="subscribe_depth" type="bool" value="true"/>
      <param name="subscribe_laserScan" type="bool" value="false"/>
      <param name="queue_size" type="int" value="30"/>
    
      <remap from="rgb/image" to="/stereo_camera/left/image_rect"/>
      <remap from="rgb/camera_info" to="/stereo_camera/left/camera_info"/>
      <remap from="depth/image" to="/stereo_camera/depth"/>
      <remap from="odom" to="/stereo_odometer/odometry"/>
      
      <param name="rgb/image_transport" type="string" value="compressed"/>
      <param name="depth/image_transport" type="string" value="compressedDepth"/>
      
      <remap from="odom" to="/stereo_odometer/odometry"/>
    </node>
    -->
    
    <node pkg="rtabmap" type="map_optimizer" name="map_optimizer" output="screen"/>
  	<node pkg="rtabmap" type="map_assembler" name="map_assembler" output="screen">
  		<remap from="mapData" to="mapData_optimized"/>
  		<param name="occupancy_grid" type="bool" value="true"/>
 	</node>
    
  </group>
  
  <!-- RVIZ -->
    
  <node pkg="rviz" type="rviz" name="rviz" args="-d $(find rtabmap)/launch/config/rgbd.rviz"/>
  <node pkg="nodelet" type="nodelet" name="standalone_nodelet"  args="manager" output="screen"/>
  <node pkg="nodelet" type="nodelet" name="data_odom_sync" args="load rtabmap/data_odom_sync standalone_nodelet">
    <remap from="rgb/image_in"       to="/stereo_camera/left/image_rect"/>
    <remap from="depth/image_in"     to="/stereo_camera/depth"/>
    <remap from="rgb/camera_info_in" to="/stereo_camera/left/camera_info"/>
    <remap from="odom_in"             to="/stereo_odometer/odometry"/>
    
    <param name="rgb/image_transport" type="string" value="compressed"/>
    <param name="depth/image_transport" type="string" value="compressedDepth"/>
    
    <remap from="rgb/image_out"       to="data_odom_sync/image"/>
    <remap from="depth/image_out"     to="data_odom_sync/depth"/>
    <remap from="rgb/camera_info_out" to="data_odom_sync/camera_info"/>
    <remap from="odom_out"            to="odom_sync"/>
    
    <param name="queue_size" type="int" value="30"/>
  </node>
  <node pkg="nodelet" type="nodelet" name="points_xyzrgb" args="load rtabmap/point_cloud_xyzrgb standalone_nodelet">
    <remap from="rgb/image"       to="data_odom_sync/image"/>
    <remap from="depth/image"     to="data_odom_sync/depth"/>
    <remap from="rgb/camera_info" to="data_odom_sync/camera_info"/>
    <remap from="cloud"           to="voxel_cloud" />
    
    <param name="queue_size" type="int" value="10"/>
    <param name="decimation" type="int" value="2"/>
  </node>
  
  <node pkg="nodelet" type="nodelet" name="points_xyz" args="load rtabmap/point_cloud_xyz standalone_nodelet">
    <remap from="depth/image"     to="data_odom_sync/depth"/>
    <remap from="depth/camera_info" to="data_odom_sync/camera_info"/>
    <remap from="cloud"           to="cloudXYZ" />
    
    <param name="queue_size" type="int" value="10"/>
    <param name="decimation" type="int" value="4"/>
    <param name="max_depth" type="double" value="4"/>
    <param name="voxel_size" type="double" value="0.05"/>
  </node>
  <node pkg="nodelet" type="nodelet" name="obstacles_detection" args="load rtabmap/obstacles_detection standalone_nodelet">
    <remap from="cloud" to="cloudXYZ"/>
  </node>
  
</launch>
